### PR TITLE
New Tutorial - Deploying an Elixir Phoenix app to GKE

### PR DESCRIPTION
**Pull request by dazuma**
This is a new tutorial on deploying an application written in Elixir using the Phoenix framework to Container Engine.

Subject matter review may be assigned to any or all of: @JustinBeckwith @bshaffer @chingor13 

- [x] Tech writing / styleguide review - @ccarpentiere 
- [x] Subject matter review - @JustinBeckwith 
- [x] Formatting review for publishing - @jmdobry

### CODE REVIEW

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    && mv /opt/release/bin/${APP_NAME} /opt/release/bin/start_server
+
+
+################################################################################
+# RUNTIME STAGE
+#
+# Creates the actual runtime image. This is based on a the Alpine Linux base
+# image, with only the minimum dependencies for running ERTS.
+
+FROM alpine:latest
```
Comment (JustinBeckwith) [inline tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile]: Nice use of the multi-stage docker syntax!  We should add a note somewhere that calls out the min version of docker required to run this.

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+root directory of the application you are building.
+
+Visit [http://localhost:8080](http://localhost:8080) to see the Phoenix
+welcome screen running locally from your Docker image.
+
+## Deploy your application
+
+Now you're ready to deploy your application to Container Engine!
+
+### Building in the cloud
```
Comment (JustinBeckwith) [inline tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: While this is cool to show, I think it's sort of a distraction, no?

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+[Kubernetes documentation](https://kubernetes.io/docs/home/) for more details.
+
+### Updating your application
+
+After you make a change to your app, redeploying is just a matter of
+building a new image and pointing your deployment to it.
+
+1.  Make a change to the app. (For example, modify the front page template.)
+
+2.  Perform a new build with a new version tag "v2":
```
Comment (JustinBeckwith) [inline tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: So this falls down pretty quick.  About the 4th time you build, you get tired of remembering how to modify the tag and v to use next.  Possibly something we could mention as an aside, but I tote this bash script with me in every k8s project I use.  
https://github.com/JustinBeckwith/cloudcats/blob/master/web/update.sh

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile (hunk 1 above)
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile]: The "before you begin" section of the tutorial does say "Install Docker 17.05 or later." Or do you mean in the Dockerfile itself?

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md (hunk 2 above)
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: How so? We need to get the image into gcr regardless. Either we run a single cloudbuild command to build and upload the image in a single go, or we build the dockerfile locally and then run a separate gcloud command to push the image.

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md (hunk 3 above)
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: Yeah I get it, it's annoying to manage tags manually. But the alternative is, as you say, provide a script or something, which complicates the tutorial and makes the process more opaque and less informative.

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,118 @@
+# Copyright 2017 Google Inc.
```
Comment (dazuma) [inline tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile]: I'm uncertain whether sample code like this should have an Apache license header (when written by a Googler). Does anyone know?

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+
+    This will run your image on a Kubernetes pod, which is the deployable unit
+    in Kubernetes. The pod opens port 8080, which is the port your Phoenix
+    application is listening on.
+
+    You may view the running pods using:
+
+        kubectl get pods
+
+2.  "Expose" the application by creating a load balancer pointing at your pod.
```
Comment (ccarpentiere) [inline tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: Double quotes don't seem necessary, here and all following instances except for the ref to "v2" at 366

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+### Deleting Container Engine resources
+
+To delete your app from Container Engine, you must remove both the load
+balancer and the Container Engine cluster.
+
+1.  Delete the service which will deallocate the load balancer:
+
+        kubectl delete service hello-web
+
+2.  The load balancer will be deleted asynchronously. Wait for that process to
```
Comment (ccarpentiere) [inline tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: Can you clarify what output the user should be looking for, or will it be obvious?

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+3.  Configure gcloud to use your cluster by default, so you don't have to
+    specify it every time for the remaining gcloud commands.
+
+        gcloud config set container/cluster hello-cluster
+
+    Replace the name if you named your cluster differently.
+
+### Deploying to the cluster
+
+A production deployment comprises two parts: your Docker container, and a
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: Actually, "comprises" is correct in this case.

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/.dockerignore [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# Ignore the following directories when doing a Docker build.
+# These directories will contain artifacts for local (not-in-Docker) builds,
+# and it is best to omit them to ensure a clean build in Docker.
+
+# Note: If you are using an umbrella app, you may need to add more directories
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/.dockerignore]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+# It includes two stages and requires Docker 17.05 or later.
+
+# To adapt this Dockerfile to your app, set the APP_NAME and PHOENIX_SUBDIR
+# build arguments, as outlined below.
+
+
+################################################################################
+# BUILD STAGE
+#
+# Builds the app, does a brunch build to build the assets, and creates a
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+## End build arguments.                                                     ##
+##############################################################################
+
+
+# Set up build environment.
+ENV MIX_ENV=prod \
+    REPLACE_OS_VARS=true \
+    TERM=xterm
+
+# We will build the app in this directory.
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    && mix local.rebar --force \
+    && mix local.hex --force
+
+# Copy the application files into /opt/app.
+COPY . .
+
+# Build the application.
+RUN mix do deps.get, deps.compile, compile
+
+# Build assets by running a brunch build and phoenix digest.
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+RUN apk update \
+    && apk --no-cache --update add bash openssl-dev
+
+# This is the runtime environment for a Phoenix app.
+# It listens on port 8080, and runs in the prod environment.
+ENV PORT=8080 \
+    MIX_ENV=prod \
+    REPLACE_OS_VARS=true
+
+# We will install the release and run in this directory.
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 13]
```diff
@@ -0,0 +1,444 @@
+---
+title: Run an Elixir Phoenix app on Kubernetes using Google Container Engine
+description: Learn how to deploy a Phoenix app to Kubernetes using Google Container Engine.
+author: dazuma
+tags: Kubernetes, Container Engine, Elixir, Phoenix
+date_published: 2017-11-01
+---
+
+This tutorial will help you get started deploying your
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 14]
```diff
@@ -0,0 +1,444 @@
...
+2. Install the `hex`, `rebar`, and `phx_new` archives.
+
+        mix local.hex
+        mix local.rebar
+        mix archive.install https://github.com/phoenixframework/archives/raw/master/phx_new.ez
+
+## Create a new app and run it locally
+
+In this section, you will create a new Phoenix app and make sure it runs. If
+you already have an app to deploy, you may use it instead.
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 15]
```diff
@@ -0,0 +1,444 @@
...
+4.  Visit [http://localhost:4000](http://localhost:4000) to see the Phoenix
+    welcome screen running locally on your workstation.
+
+## Enable releases with Distillery
+
+Releases are the preferred way to package Elixir (and Erlang) applications for
+deployment. You will configure the Distillery tool to create releases for your
+app.
+
+**Note:** If you already have Distillery set up for your application, you may
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 16]
```diff
@@ -0,0 +1,444 @@
...
+          check_origin: false,
+          server: true,
+          root: ".",
+          cache_static_manifest: "priv/static/cache_manifest.json"
+
+### Testing a release
+
+Now you can create a release to test out your configuration.
+
+1.  First build the application for production:
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: Removed "first"

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 17]
```diff
@@ -0,0 +1,444 @@
...
+
+        MIX_ENV=prod mix deps.compile, compile
+        cd assets
+        npm install
+        ./node_modules/brunch/bin/brunch build -p
+        cd ..
+        mix phx.digest
+
+    Remember that if your app is an umbrella app, the assets directory may be
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 18]
```diff
@@ -0,0 +1,444 @@
...
+
+The next step is to produce a Docker image that builds and runs your
+application in a Docker container. You will define this image using a
+Dockerfile.
+
+### Creating the Dockerfile
+
+Various considerations go into designing a good Docker image. The Dockerfile
+used by this tutorial will build a release and run it with Alpine Linux.
+If you are experienced with Docker, you may choose to customize your image.
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: Used "can"

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 19]
```diff
@@ -0,0 +1,444 @@
...
+
+1.  Create a file called `Dockerfile` in your `hello` directory. Copy the
+    following content into it. Alternately, you may
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 20]
```diff
@@ -0,0 +1,444 @@
...
+        assets/node_modules
+
+    **Note:** if your app is an umbrella app, you may need to adjust the paths
+    to include the build, deps, and node_modules directories of the constituent
+    apps. In general, you want Docker to ignore artifacts that come from your
+    development environment, so it can perform clean builds.
+
+### Testing the Dockerfile
+
+Now build the image locally and test running your application from the image:
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: Removed "now"

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 21]
```diff
@@ -0,0 +1,444 @@
...
+
+### Building in the cloud
+
+To deploy the app, you will use the
+[Google Cloud Container Build](https://cloud.google.com/container-builder/)
+service to build your Docker image in the cloud and store the resulting Docker
+image in your project in the
+[Google Cloud Container Registry](https://cloud.google.com/container-registry/).
+
+Execute the following to run the build:
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 22]
```diff
@@ -0,0 +1,444 @@
...
+    name is "hello-cluster".
+
+2.  Create the cluster.
+
+        gcloud container clusters create hello-cluster --num-nodes=2
+
+    This command creates a cluster of two machines. You may choose a different
+    size, but two is a good starting point.
+
+    It may take several mintues for the cluster to be created. You can check
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: whoops. fixed.

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 23]
```diff
@@ -0,0 +1,444 @@
...
+    the cloud console at http://cloud.google.com/console, under the Container
+    Engine section, to see that your cluster is running. You will also be able
+    to see the individual running VMs under the Compute Engine section. Note
+    that once the cluster is running, you will be charged for the VM usage.
+
+3.  Configure gcloud to use your cluster by default, so you don't have to
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md [hunk 24]
```diff
@@ -0,0 +1,444 @@
...
+
+### Deploying to the cluster
+
+A production deployment comprises two parts: your Docker container, and a
+front-end load balancer (which also provides a public IP address.)
+
+We'll assume that you built the image to `gcr.io/${PROJECT_ID}/hello:v1` and
+you've created the Kubernetes cluster as described above.
+
+1.  Create a deployment
```
[the remark this answers is not in the document]
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md (hunk 5 above)
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: done

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md (hunk 6 above)
Comment (dazuma) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: I think it should be obvious, but I added a sentence.

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile (hunk 1 above)
Comment (JustinBeckwith) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile]: I noticed that further down.  Disregard :)

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md (hunk 3 above)
Comment (JustinBeckwith) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/index.md]: Fair point.  I'm ok with this as it is :)

File: tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile (hunk 4 above)
Comment (jmdobry) [reply tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile]: The current license is correct. Code in this repository is Apache-2.0 licensed.

**Comment by jmdobry** [comment]
Published: https://cloud.google.com/community/tutorials/elixir-phoenix-on-kubernetes-google-container-engine

**Pull request merged**